File: UNIT-01/orchestrator/src/clients/indexer.rs
use super::uds::UDSClient;
use serde::Deserialize;
use serde_json::json;

pub struct IndexerClient {
    client: UDSClient,
}

impl IndexerClient {
    pub fn new() -> Self {
        Self {
            client: UDSClient::new("/tmp/ruthen/indexer.sock"),
        }
    }

    fn auth_params() -> serde_json::Value {
        json!({"token": "uds-internal-trust"})
    }

    fn merge(mut base: serde_json::Value, extra: serde_json::Value) -> serde_json::Value {
        if let (Some(b), Some(e)) = (base.as_object_mut(), extra.as_object()) {
            for (k, v) in e {
                b.insert(k.clone(), v.clone());
            }
        }
        base
    }

    pub async fn search(&self, query: &str) -> Result<Vec<FileRecord>, String> {
        #[derive(Deserialize)]
        struct SearchWrapper(Vec<FileRecord>);

        let p = Self::merge(json!({"query": query}), Self::auth_params());
        let res: Vec<FileRecord> = self.client.call("search", p).await?;
        Ok(res)
    }

    pub async fn read(&self, path: &str) -> Result<String, String> {
        #[derive(Deserialize)]
        struct ReadResult {
            content: String,
        }

        let res: ReadResult = self
            .client
            .call(
                "read",
                Self::merge(json!({"path": path}), Self::auth_params()),
            )
            .await?;
        Ok(res.content)
    }

    pub async fn list(&self, path: &str) -> Result<ListResult, String> {
        let res: ListResult = self
            .client
            .call(
                "ls",
                Self::merge(json!({"path": path}), Self::auth_params()),
            )
            .await?;
        Ok(res)
    }

    pub async fn write(&self, path: &str, content: &str) -> Result<(), String> {
        self.client
            .call::<serde_json::Value>(
                "write",
                Self::merge(
                    json!({"path": path, "content": content}),
                    Self::auth_params(),
                ),
            )
            .await?;
        Ok(())
    }

    pub async fn patch(
        &self,
        path: &str,
        target: &str,
        replacement: &str,
    ) -> Result<String, String> {
        #[derive(Deserialize)]
        struct PatchResult {
            status: String,
        }
        let res: PatchResult = self
            .client
            .call(
                "patch",
                Self::merge(
                    json!({"path": path, "target": target, "replacement": replacement}),
                    Self::auth_params(),
                ),
            )
            .await?;
        Ok(res.status)
    }

    pub async fn delete(&self, path: &str) -> Result<String, String> {
        #[derive(Deserialize)]
        struct DeleteResult {
            status: String,
        }
        let res: DeleteResult = self
            .client
            .call(
                "delete",
                Self::merge(json!({"path": path}), Self::auth_params()),
            )
            .await?;
        Ok(res.status)
    }

    pub async fn rollback(&self) -> Result<String, String> {
        #[derive(Deserialize)]
        struct RbResult {
            status: String,
        }
        let res: RbResult = self
            .client
            .call("rollback", Self::merge(json!({}), Self::auth_params()))
            .await?;
        Ok(res.status)
    }

    pub async fn get_project_map(&self, path: &str) -> Result<String, String> {
        #[derive(Deserialize)]
        struct MapResult {
            map: String,
        }
        let res: MapResult = self
            .client
            .call(
                "project_map",
                Self::merge(json!({"path": path}), Self::auth_params()),
            )
            .await?;
        Ok(res.map)
    }

    pub async fn glob(&self, pattern: &str, base: &str) -> Result<Vec<String>, String> {
        #[derive(Deserialize)]
        struct GlobResult {
            files: Vec<String>,
        }
        let res: GlobResult = self
            .client
            .call(
                "glob",
                Self::merge(
                    json!({"pattern": pattern, "base": base}),
                    Self::auth_params(),
                ),
            )
            .await?;
        Ok(res.files)
    }

    pub async fn find(&self, name: &str, root: &str) -> Result<Vec<String>, String> {
        #[derive(Deserialize)]
        struct FindResult {
            files: Vec<String>,
        }
        let res: FindResult = self
            .client
            .call(
                "find",
                Self::merge(json!({"name": name, "root": root}), Self::auth_params()),
            )
            .await?;
        Ok(res.files)
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct FileRecord {
    pub path: String,
    #[serde(default)]
    pub relative_path: String,
    #[serde(default)]
    pub language: String,
    pub size_bytes: u64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ListEntry {
    pub name: String,
    #[serde(rename = "type")]
    pub entry_type: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ListResult {
    pub entries: Vec<ListEntry>,
}

File: UNIT-01/cli/src/daemons/indexer.ts
// ── Indexer daemon client ─────────────────────────────────────────────

import { rpcOnce, DaemonError } from './socket.js'

export const INDEXER_SOCKET = '/tmp/ruthen/indexer.sock'

export interface SearchResult {
  path: string
  content: string
  score: number
  language?: string
  line?: number
  span?: [number, number]
}

export interface GlobResult { files: string[] }
export interface FindResult { files: string[] }
export interface FileInfo {
  size: number
  is_dir: boolean
  modified: number
}
export interface ReadResult { content: string }
export interface WriteResult { status: string }
export interface PatchResult { status: string }
export interface ShadowEntry { path_hash: string; original_path: string }
export interface ShadowListResult { entries: ShadowEntry[]; count: number }
export interface DependentsResult { dependents: string[] }
export interface DependenciesResult { dependencies: string[] }
export interface IndexDepsResult { indexed: number; nodes: number }
export interface ImpactResult { impact: string }

export class IndexerClient {
  socketPath: string
  constructor(socketPath = INDEXER_SOCKET) {
    this.socketPath = socketPath
  }

  status() {
    return rpcOnce<{ status: string }>(this.socketPath, 'status', {})
  }
  stop() {
    return rpcOnce<{ status: string }>(this.socketPath, 'stop', {})
  }
  search(query: string, opts: { limit?: number; lang?: string; path?: string } = {}) {
    return rpcOnce<{ results: SearchResult[]; count: number }>(this.socketPath, 'search', {
      query,
      limit: opts.limit ?? 20,
      ...(opts.lang ? { lang: opts.lang } : {}),
      ...(opts.path ? { path: opts.path } : {}),
    })
  }
  semanticSearch(query: string, limit = 10) {
    return rpcOnce<{ results: SearchResult[]; count: number }>(this.socketPath, 'semantic_search', { query, limit })
  }
  glob(pattern: string, base = '.') {
    return rpcOnce<GlobResult>(this.socketPath, 'glob', { pattern, base })
  }
  find(name: string, root = '.') {
    return rpcOnce<FindResult>(this.socketPath, 'find', { name, root })
  }
  fileInfo(path: string) {
    return rpcOnce<FileInfo>(this.socketPath, 'file_info', { path })
  }
  read(path: string) {
    return rpcOnce<ReadResult>(this.socketPath, 'read', { path })
  }
  write(path: string, content: string) {
    return rpcOnce<WriteResult>(this.socketPath, 'write', { path, content })
  }
  patch(path: string, target: string, replacement: string) {
    return rpcOnce<PatchResult>(this.socketPath, 'patch', { path, target, replacement })
  }
  dependents(path: string) {
    return rpcOnce<DependentsResult>(this.socketPath, 'dependents', { path })
  }
  dependencies(path: string) {
    return rpcOnce<DependenciesResult>(this.socketPath, 'dependencies', { path })
  }
  transitiveDependents(path: string) {
    return rpcOnce<DependentsResult>(this.socketPath, 'transitive_dependents', { path })
  }
  findExport(symbol: string) {
    return rpcOnce<{ files: string[] }>(this.socketPath, 'find_export', { symbol })
  }
  impact(path: string) {
    return rpcOnce<ImpactResult>(this.socketPath, 'impact', { path })
  }
  indexDeps(path = '.') {
    return rpcOnce<IndexDepsResult>(this.socketPath, 'index_deps', { path })
  }
  shadowList() {
    return rpcOnce<ShadowListResult>(this.socketPath, 'shadow_list', {})
  }
  rollback() {
    return rpcOnce<{ status: string }>(this.socketPath, 'rollback', {})
  }

  // Helper: read with line window
  async readLines(path: string, start: number, end: number): Promise<{ content: string; total: number }> {
    const r = await this.read(path)
    const lines = r.content.split('\n')
    const from = Math.max(0, start - 1)
    const to = Math.min(lines.length, end)
    return { content: lines.slice(from, to).join('\n'), total: lines.length }
  }

  isAlive(): boolean {
    try {
      // Fast reachability check
      return Bun.file(this.socketPath).size > 0
    } catch {
      return false
    }
  }
}
